desc/basis.py

```diff
@@ -1413,6 +1413,424 @@ def zernike_radial_poly(r, l, m, dr=0, exact="auto"):
+        P_n = jacobi_poly_single(r_jacobi, N, alpha, 0, P_n1, P_n2)
+        dP_n = jacobi_poly_single(r_jacobi, N - 1, alpha + 1, 1, dP_n1, dP_n2)
+        ddP_n = jacobi_poly_single(r_jacobi, N - 2, alpha + 2, 2, ddP_n1, ddP_n2)
+        dddP_n = jacobi_poly_single(r_jacobi, N - 3, alpha + 3, 3, dddP_n1, dddP_n2)
+        ddddP_n = jacobi_poly_single(r_jacobi, N - 4, alpha + 4, 4, ddddP_n1, ddddP_n2)
+
+        # Find the index corresponding to the original array
+        index = jnp.where(
+            jnp.logical_and(m == alpha, n == N),
+            jnp.arange(m.size),
+            0,
+        )
+        # jnp.where() will result in 1D array with 1 value which
+        # is not 0. Sum it to find the index
+        idx = jnp.sum(index)
+
+        coef_1 = gammaln(alpha + N + 2) - jnp.log(2) - gammaln(alpha + N + 1)
+        coef_2 = gammaln(alpha + N + 3) - 2 * jnp.log(2) - gammaln(alpha + N + 1)
+        coef_3 = gammaln(alpha + N + 4) - 3 * jnp.log(2) - gammaln(alpha + N + 1)
+        coef_4 = gammaln(alpha + N + 5) - 4 * jnp.log(2) - gammaln(alpha + N + 1)
+        coef_1 = jnp.exp(coef_1)
+        coef_2 = jnp.exp(coef_2)
+        coef_3 = jnp.exp(coef_3)
+        coef_4 = jnp.exp(coef_4)
+
+        result = (
+            (alpha - 3)
+            * (alpha - 2)
+            * (alpha - 1)
+            * alpha
+            * r ** jnp.maximum(alpha - 4, 0)
+            * P_n
+            - coef_1
+            * 8
+            * alpha
+            * (2 * alpha**2 - 3 * alpha + 1)
+            * r ** jnp.maximum(alpha - 2, 0)
+            * dP_n
+            + coef_2 * 48 * (2 * alpha**2 + 2 * alpha + 1) * r**alpha * ddP_n
+            - coef_3 * 128 * (2 * alpha + 3) * r ** (alpha + 2) * dddP_n
+            + coef_4 * 256 * r ** (alpha + 4) * ddddP_n
+        )
+        out = out.at[:, idx].set((-1) ** N * result)
+
+        P_n2 = jnp.where(N >= 2, P_n1, P_n2)
+        P_n1 = jnp.where(N >= 2, P_n, P_n1)
+        dP_n2 = jnp.where(N >= 3, dP_n1, dP_n2)
+        dP_n1 = jnp.where(N >= 3, dP_n, dP_n1)
+        ddP_n2 = jnp.where(N >= 4, ddP_n1, ddP_n2)
+        ddP_n1 = jnp.where(N >= 4, ddP_n, ddP_n1)
+        dddP_n2 = jnp.where(N >= 5, dddP_n1, dddP_n2)
+        dddP_n1 = jnp.where(N >= 5, dddP_n, dddP_n1)
+        ddddP_n2 = jnp.where(N >= 6, ddddP_n1, ddddP_n2)
+        ddddP_n1 = jnp.where(N >= 6, ddddP_n, ddddP_n1)
+
+        return (
+            r_jacobi,
+            alpha,
+            r,
+            out,
+            (
+                P_n1,
+                P_n2,
+                dP_n1,
+                dP_n2,
+                ddP_n1,
+                ddP_n2,
+                dddP_n1,
+                dddP_n2,
+                ddddP_n1,
+                ddddP_n2,
+            ),
+            m,
+            n,
+        )
+
+    def body(alpha, args):
+        r, r_jacobi, L_max, m, n, out = args
+        # Maximum possible value for n for loop bound
+        N_max = (L_max - alpha) // 2
```

Comment:
Are we sure this only depends on Lmax and not Mmax?

Reply:
The purpose of body function is to loop over every m value from 0 to Mmax, assuming that we have every mode. However, again by assuming we have every mode, the highest value corresponding to any m is Lmax. I am not very familiar with other basis uses, so I hoped my assumptions are valid, but if not I can chane Lmax with jnp.max(l) kind of thing